File: RomanConverter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace RomanConverter
{
    public class RomanConverter
    {
        /// <summary>
        /// Converte um número inteiro (1-3999) para algarismos romanos.
        /// </summary>
        /// <param name="num">Número inteiro a ser convertido</param>
        /// <returns>Representação em algarismos romanos</returns>
        /// <exception cref="ArgumentOutOfRangeException">Quando o número não está entre 1 e 3999</exception>
        public static string ToRoman(int num)
        {
            if (num < 1 || num > 3999)
            {
                throw new ArgumentOutOfRangeException(
                    nameof(num),
                    "Número deve ser um inteiro entre 1 e 3999"
                );
            }

            // Mapeamento de valores decimais para romanos em ordem decrescente
            var valores = new List<(int valor, string simbolo)>
            {
                (1000, "M"),
                (900, "CM"),
                (500, "D"),
                (400, "CD"),
                (100, "C"),
                (90, "XC"),
                (50, "L"),
                (40, "XL"),
                (10, "X"),
                (9, "IX"),
                (5, "V"),
                (4, "IV"),
                (1, "I"),
            };

            string resultado = "";

            foreach (var (valor, simbolo) in valores)
            {
                while (num >= valor)
                {
                    resultado += simbolo;
                    num -= valor;
                }
            }

            return resultado;
        }

        /// <summary>
        /// Lê o arquivo de teste e verifica se o método ToRoman está funcionando corretamente.
        /// </summary>
        public static void TestarComArquivo()
        {
            string arquivoTeste = "decimais_romanos_1_a_3999.txt";

            try
            {
                string[] linhas = File.ReadAllLines(arquivoTeste);

                Console.WriteLine($"Testando {linhas.Length} conversões...");
                Console.WriteLine(new string('-', 50));

                int erros = 0;
                int sucessos = 0;

                for (int i = 0; i < linhas.Length; i++)
                {
                    string linha = linhas[i].Trim();
                    if (string.IsNullOrEmpty(linha))
                        continue;

                    string[] partes = linha.Split(' ', 2);
                    if (partes.Length != 2)
                    {
                        Console.WriteLine($"Linha {i + 1} inválida: {linha}");
                        continue;
                    }

                    try
                    {
                        int numero = int.Parse(partes[0]);
                        string romanoEsperado = partes[1];
                        string romanoCalculado = ToRoman(numero);

                        if (romanoCalculado == romanoEsperado)
                        {
                            sucessos++;
                            // Mostra os primeiros 10 e depois a cada 100
                            if (i < 10 || (i + 1) % 100 == 0)
                            {
                                Console.WriteLine($"✓ {numero} → {romanoCalculado}");
                            }
                        }
                        else
                        {
                            erros++;
                            Console.WriteLine(
                                $"✗ {numero}: esperado '{romanoEsperado}', obtido '{romanoCalculado}'"
                            );
                        }
                    }
                    catch (Exception e)
                    {
                        erros++;
                        Console.WriteLine($"Erro na linha {i + 1}: {e.Message}");
                    }
                }

                Console.WriteLine(new string('-', 50));
                Console.WriteLine("Resultados do teste:");
                Console.WriteLine($"✓ Sucessos: {sucessos}");
                Console.WriteLine($"✗ Erros: {erros}");
                Console.WriteLine($"Total: {sucessos + erros}");

                if (erros == 0)
                {
                    Console.WriteLine(
                        "\n🎉 Todos os testes passaram! O método ToRoman está funcionando corretamente."
                    );
                }
                else
                {
                    Console.WriteLine($"\n⚠️  {erros} testes falharam. Verifique a implementação.");
                }
            }
            catch (FileNotFoundException)
            {
                Console.WriteLine($"Arquivo '{arquivoTeste}' não encontrado!");
            }
            catch (Exception e)
            {
                Console.WriteLine($"Erro ao processar arquivo: {e.Message}");
            }
        }

        /// <summary>
        /// Método principal para executar o conversor e os testes.
        /// </summary>
        public static void Main(string[] args)
        {
            Console.WriteLine("=== Conversor de Números para Algarismos Romanos ===\n");

            // Executar testes com o arquivo
            TestarComArquivo();
        }
    }
}

File: roman_converter.py
def to_roman(num):
    """
    Converte um número inteiro (1-3999) para algarismos romanos.

    Args:
        num (int): Número inteiro a ser convertido

    Returns:
        str: Representação em algarismos romanos
    """
    if not isinstance(num, int) or num < 1 or num > 3999:
        raise ValueError("Número deve ser um inteiro entre 1 e 3999")

    # Mapeamento de valores decimais para romanos em ordem decrescente
    valores = [
        (1000, "M"),
        (900, "CM"),
        (500, "D"),
        (400, "CD"),
        (100, "C"),
        (90, "XC"),
        (50, "L"),
        (40, "XL"),
        (10, "X"),
        (9, "IX"),
        (5, "V"),
        (4, "IV"),
        (1, "I"),
    ]

    resultado = ""

    for valor, simbolo in valores:
        while num >= valor:
            resultado += simbolo
            num -= valor

    return resultado


def testar_com_arquivo():
    """
    Lê o arquivo de teste e verifica se o método to_roman está funcionando corretamente.
    """
    arquivo_teste = "decimais_romanos_1_a_3999.txt"

    try:
        with open(arquivo_teste, "r", encoding="utf-8") as arquivo:
            linhas = arquivo.readlines()

        print(f"Testando {len(linhas)} conversões...")
        print("-" * 50)

        erros = 0
        sucessos = 0

        for i, linha in enumerate(linhas, 1):
            linha = linha.strip()
            if not linha:
                continue

            partes = linha.split(" ", 1)
            if len(partes) != 2:
                print(f"Linha {i} inválida: {linha}")
                continue

            try:
                numero = int(partes[0])
                romano_esperado = partes[1]
                romano_calculado = to_roman(numero)

                if romano_calculado == romano_esperado:
                    sucessos += 1
                    if (
                        i <= 10 or i % 100 == 0
                    ):  # Mostra os primeiros 10 e depois a cada 100
                        print(f"✓ {numero} → {romano_calculado}")
                else:
                    erros += 1
                    print(
                        f"✗ {numero}: esperado '{romano_esperado}', obtido '{romano_calculado}'"
                    )

            except ValueError as e:
                erros += 1
                print(f"Erro na linha {i}: {e}")

        print("-" * 50)
        print(f"Resultados do teste:")
        print(f"✓ Sucessos: {sucessos}")
        print(f"✗ Erros: {erros}")
        print(f"Total: {sucessos + erros}")

        if erros == 0:
            print(
                "\n🎉 Todos os testes passaram! O método to_roman está funcionando corretamente."
            )
        else:
            print(f"\n⚠️  {erros} testes falharam. Verifique a implementação.")

    except FileNotFoundError:
        print(f"Arquivo '{arquivo_teste}' não encontrado!")
    except Exception as e:
        print(f"Erro ao processar arquivo: {e}")



if __name__ == "__main__":
    print("=== Conversor de Números para Algarismos Romanos ===\n")

    # Executar testes com o arquivo
    testar_com_arquivo()
